### PR TITLE
Improve sourcegraph release process docs

### LINKED ISSUE #30420: Improve release process documentation
## The problem

The current documented [release process] on the handbook doesn't really tell you how it works. We need to update the process section to reflect how does the release process work from start to end

## Goal

At a glance, I can tell what steps are necessarily required to make a minor or patch release. It can be a link to some other resources, and that's okay. The point is having the process explained in a centralized place instead of random GitHub issues or PRs flying around and linked to each other.

## Non-goal

Revamp the current [release process] to make it easier or fully automated. We acknowledge the fact that the process is tedious and hard to follow, but this belongs to another conversation.

[release process]: https://handbook.sourcegraph.com/departments/product-engineering/engineering/process/releases#release-process

### PR DESCRIPTION
**Pull request by michaellzc**
Fix https://github.com/sourcegraph/sourcegraph/issues/30420

### CODE REVIEW

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (caugustus-sourcegraph) [inline content/departments/product-engineering/engineering/process/releases/index.md]: Maybe we should add another section detailing how to decide whether to accept the patch - who gets to decide? Who needs to be part of the conversation, and who gets notified? What's the timeframe?

I don't think we have good answers - pretty much the release captain says 'yes' and immediately works on the patch. CS generally gets looped in by either the dev requesting the patch or the release captain, since a patch almost always causes additional work for them.

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (caugustus-sourcegraph) [inline content/departments/product-engineering/engineering/process/releases/index.md]:
```suggestion
### 1) Kickstart a patch release using the release tool
```

Comment (caugustus-sourcegraph) [reply content/departments/product-engineering/engineering/process/releases/index.md]: Small suggestion - I'm hoping that we can make it so that a release captain can scan the headings to get a reminder of what to do, and only dive into the details if necessary.

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (caugustus-sourcegraph) [inline content/departments/product-engineering/engineering/process/releases/index.md]: I wonder how valuable it is to include these steps - they're already included in the tracking issue itself, and it means that if we change that issue this list can get out of sync.

Maybe instead, we can just summarize the intent of the process? Build a release candidate, verify it passes CI, publish final image and update documentation/references.

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (caugustus-sourcegraph) [inline content/departments/product-engineering/engineering/process/releases/index.md]: It might be better to reword this as an active sentence - confirm that CI checks have passed and all manual steps detailed in the PR have been addressed

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (caugustus-sourcegraph) [inline content/departments/product-engineering/engineering/process/releases/index.md:121]: Totally silly, but in my mind "Minor release" should go before "Patch release" in the doc 😄

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (caugustus-sourcegraph) [inline content/departments/product-engineering/engineering/process/releases/index.md]: Again, I think it could be worthwhile to replace the specific steps with a summary of the intent.

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (caugustus-sourcegraph) [inline content/departments/product-engineering/engineering/process/releases/index.md]: There are a few typos / grammatical errors in the doc - probably worth running a spellchecker against your changes.

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (bobheadxi) [inline content/departments/product-engineering/engineering/process/releases/index.md:57]: Why remove this link? At the very least it should point to the source code

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (bobheadxi) [reply content/departments/product-engineering/engineering/process/releases/index.md]: +1 - the release issue documents this, and the issue template is already available in the handbook

**Review by bobheadxi — COMMENTED** [review]
Out of curiosity, are the release issue templates insufficient documentation? Maybe we should improve those instead, so that release captains automatically have fleshed out resources on hand and don't have to check in multiple places. Chiming in because i've worked extensively on this process in the past

https://handbook.sourcegraph.com/departments/product-engineering/engineering/process/releases/release_issue_template

https://handbook.sourcegraph.com/departments/product-engineering/engineering/process/releases/patch_release_issue_template

The release process could be updated to more clearly indicate these templates are the release process

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (kevinwojo) [inline content/departments/product-engineering/engineering/process/releases/index.md]: Not a definite thing, but my question from the outside perspective, is "Why do I need to wait for CI to pass"? I'm wondering if it's worth putting that type of information in here? 🤷

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (kevinwojo) [inline content/departments/product-engineering/engineering/process/releases/index.md:177]: Is this true? I think it's actually more nuanced than this. Our customers usually don't have the capacity to upgrade continuously. It can take a customer months (or years) to get to the next incremental release depending on their internal processes. A lot of our customers have rigorous change controls in-place.

I agree that our automated testing could use some work, but definitely think the issue at hand is a little more complicated 😆 

I know you didn't make this change, just saw it and felt something about this line!

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (michaellzc) [reply content/departments/product-engineering/engineering/process/releases/index.md]: LOL this is my impression too, like the Release Captain has to wait for multiple CI builds and each takes 30+min to run

- the cherry-pick one on the release branch
- release candidate tag build
- final release tag build

so much time is wasted. 

Maybe we can only wait for one of the CI runs to pass?

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (michaellzc) [reply content/departments/product-engineering/engineering/process/releases/index.md:57]: > Why remove this link? At the very least it should point to the source code

It's some fancy markdown anchor link syntax, we can reuse the same anchor link to keep things DRY. 

https://github.com/sourcegraph/handbook/pull/2059/files#diff-23bdf5adea38c4be7b98df79bb9f1b1959d5380a25ad3e269abccb12f91fb0f2R249

```md
[sourcegraph release tool] is great

[sourcegraph release tool] is awesome

[sourcegraph release tool]: ../../tools/release.md
```

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (michaellzc) [reply content/departments/product-engineering/engineering/process/releases/index.md]: good point, also the handbook and the issue template could eventually get out of sync.

I am removing all specific steps here and move some nice to have notes (the one about how to merge the batch change) into the release template.

**Comment by michaellzc** [comment]
> Out of curiosity, are the release issue templates insufficient documentation? Maybe we should improve those instead, so that release captains automatically have fleshed out resources on hand and don't have to check in multiple places. Chiming in because i've worked extensively on this process in the past
> 
> [handbook.sourcegraph.com/departments/product-engineering/engineering/process/releases/release_issue_template](https://handbook.sourcegraph.com/departments/product-engineering/engineering/process/releases/release_issue_template)
> 
> [handbook.sourcegraph.com/departments/product-engineering/engineering/process/releases/patch_release_issue_template](https://handbook.sourcegraph.com/departments/product-engineering/engineering/process/releases/patch_release_issue_template)
> 
> The release process could be updated to more clearly indicate these templates are the release process

This PR is trying to address https://github.com/sourcegraph/sourcegraph/issues/30420

The intention of this PR is not to replace the existing stuff (release templates, etc), we are hoping this can serve as a one-page summary of the overall release process. Anyone can just look at this doc can tell, this is how a release is done instead of going to GitHub issues and looking up existing tracking issues in order to understand how things work.

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (michaellzc) [reply content/departments/product-engineering/engineering/process/releases/index.md:121]: lol, it makes sense. I just happened to write up the patch release first. I'll move them around.

**Comment by bobheadxi** [comment]
> The intention of this PR is not to replace the existing stuff (release templates, etc), we are hoping this can serve as a one-page summary of the overall release process. Anyone can just look at this doc can tell, this is how a release is done instead of going to GitHub issues and looking up existing tracking issues in order to understand how things work.

I am very concerned that this is highly duplicative of the tracking issue, and am a strong proponent of keeping docs centralized in one single document. From what I've seen docs get outdated and misaligned frequently and easily, and that drift just leaves processes and guidelines in a confusing state.

From the issue:

> At a glance, I can tell what steps are necessarily required to make a minor or patch release. It can be a link to some other resources, and that's okay. The point is having the process explained in a centralized place instead of random GitHub issues or PRs flying around and linked to each other.

I believe this is solved by centralizing all documentation in the issue template - right now the process is explained in this one central place - and not by splitting documentation into multiple pages

In other words - for a specific release, the _generated_ issue gives an accurate, up-to-date step-by-step framework for that specific release. For overall processes, the handbook (as always) is the source of truth, in which case the _template_ issue gives the most updated processes. I do not believe we are asking captains to follow random GitHub issues to find the appropriate process - there is a very specific issue, and additional specific action items are closely linked from issues with a variety of tooling to help you track progress.

Also from the issue:

> The current documented release process on the handbook doesn't really tell you how it works. We need to update the process section to reflect how does the release process work from start to end

I definitely agree - but the solution is to make the release templates more prominent, and not to reword the template instructions into another page. Sort of like [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself#:~:text=The%20DRY%20principle%20is%20stated,their%20book%20The%20Pragmatic%20Programmer.), but for docs 😄

**Comment by bobheadxi** [comment]
Also from the issue:

> The current documented release process on the handbook doesn't really tell you how it works. We need to update the process section to reflect how does the release process work from start to end

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (bobheadxi) [inline content/departments/product-engineering/engineering/process/releases/index.md:152]: We should just link to the template here

Comment (bobheadxi) [inline content/departments/product-engineering/engineering/process/releases/index.md]: We should just link to the template here, instead of a potentially outdated example

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (bobheadxi) [inline content/departments/product-engineering/engineering/process/releases/index.md]: Similar: We should just link to the template here, instead of a potentially outdated example

**Review by bobheadxi — COMMENTED** [review]
Looks like I wrote [my previous comment](https://github.com/sourcegraph/handbook/pull/2059#issuecomment-1027129489) a bit too soon - I like the direction these changes are going in now 😁  Thank you for taking the time to do this update!

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (caugustus-sourcegraph) [inline content/departments/product-engineering/engineering/process/releases/index.md]:
```suggestion
> As of Jan 31, 2022, the [Delivery] team owns the release process. The role of Release Captain is referring to a member of the [Delivery] team. However, this may change in the future when we transition to a rotation-based Release Captain.
```
I kinda don't want to fix this, because the original sentence makes me imagine something like this:
hamster

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (michaellzc) [reply content/departments/product-engineering/engineering/process/releases/index.md:177]: I am a bit confused.

Are you saying we do have some `automated testing and update infrastructure` for releases? 

About enabling customers to continuously update, the biggest blocker is 

> We don't think customers would be comfortable with a continuously updated service running on their own infrastructure, for security and stability reasons.

If we throw this away, and we do want to CD for on-prem. I saw Dan is talking about giving [Replicated](https://www.replicated.com/) another try. This could potentially allow us to have an automated release process for customers.

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (caugustus-sourcegraph) [inline content/departments/product-engineering/engineering/process/releases/index.md]:
```suggestion
The Release Captain should review and follow the instruction in the release tracking release for the next steps. At a high level, it includes the following steps:
```

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (caugustus-sourcegraph) [inline content/departments/product-engineering/engineering/process/releases/index.md]:
```suggestion
Follow the instruction from post-release in the release tracking issue to wrap up the release and schedule the next minor release.
```

File: content/departments/product-engineering/engineering/process/releases/index.md
Comment (michaellzc) [reply content/departments/product-engineering/engineering/process/releases/index.md]: 😂English is hard

**Review by caugustus-sourcegraph — APPROVED** [review]

**Pull request merged**
